### ISSUE TITLE
new rule - enforce that `self` parameter on methods is a supertype of its class

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ty#3666]
```py
class Foo:
    def fn(asdf: int): ...
```
both [pyright](https://basedpyright.com/?typeCheckingMode=all&code=MYGwhgzhAEBiD28BcAoa7oBMCmAzauAdgBQTYi5LQCWhALgJRUB0rQA) and [pyrefly](https://pyrefly.org/sandbox/?project=N4IgZglgNgpgziAXKOBDAdgEwEYHsAeAdAA4CeSIAxlKnHAAQBiuuiAOuvV-ZjGPWHQAKWpjCJ6EdABcAlBMKKQAGhBkATnyilC03AFsoFMtIAWudAFoAbjHVwIF%2BgF56bEAGZCARgAs7jmJNOBhpFzcQOGl1CEppAPQAbTt1XHsAXQ5eIJhKVGkYTHD3FLT3N3QIfWIoWIhpS1RsKPVUOMtqWgZXEvVU9QSqmrqGobSwnpBSgZAKodrKeu1LXkh0QsbpaIhsAFcC4qm%2BspAOfQg6KQBzSzhcXfVKGEPphPQLSxh8U1RdqIhbJZ9PlKKYXscZhx3g1NABHXYQTSYSwAaxgpEalCedHB-QSuGIMCsu3QxDaKOuuJOFSRJMwGAaeSiVJmNMKdIZHQsmHqjk4k1esw4JPW2LQ6gxlAMZJid35EUFFRFuXg4ox0lIhK56Fs9j5LPKwvQmjaP2wsANQvQJJNoKasCBINMHVozwFEMN1uN8FwUGs9pgKzpdkaUCglqVMk1GzG6gmCo9VpJfxjV3emgjHH9MQwTyBF2B0lBmcqOtQtWRvClrT06ktIAAvqo2tIATBGNAYBQ0Fg8EQyI2gA) report an error here:

```
Type of parameter "self" must be a supertype of its class "Foo"  (reportGeneralTypeIssues)
```
```
ERROR sandbox.py:2:9-11: `fn` method self type `int` is not a superclass of class `Foo` [invalid-annotation]
```
